// @rxweb/angular-router/esm2015/directives/rx-authorize.directive.js
import * as tslib_1 from "tslib";
import { Injector, Directive, ViewContainerRef, TemplateRef, Input } from "@angular/core";
import { NgIfContext } from "@angular/common";
import { routeContainer } from "../core";
import { componentInstanceProvider } from "../core/component-instance-provider";
let RxAuthorizeDirective = class RxAuthorizeDirective {
    constructor(viewContainerRef, templateRef, injector) {
        this.viewContainerRef = viewContainerRef;
        this.templateRef = templateRef;
        this.injector = injector;
        this.viewRef = null;
        this._context = new NgIfContext();
    }
    set component(value) {
        if (Array.isArray(value)) {
            var result = true;
            this._components = value;
            this.checkAuth(0);
        }
        else if (value)
            this.checkAccess(value);
    }
    checkAuth(index) {
        if (this._components.length > index) {
            this.checkAccess(this._components[index], index);
        }
        else {
            this.updateView(true);
        }
    }
    checkAccess(value, currentIndex = 0) {
        var authorizeModel = routeContainer.get().authorization;
        let component = routeContainer.getModelDecorator(value, "access");
        if (authorizeModel && component) {
            var authorize = this.injector.get(authorizeModel);
            var authorizeConfig = componentInstanceProvider.getAuthorizeConfig();
            var result = authorize.authorizeChildren(component.functions, authorizeConfig);
            if (typeof result === "boolean") {
                if (this._components && this._components.length > 0 && result) {
                    let index = currentIndex + 1;
                    this.checkAuth(index);
                }
                else
                    this.updateView(result);
            }
            else
                result.then(t => {
                    if (t) {
                        let index = currentIndex + 1;
                        this.checkAuth(index);
                    }
                    else
                        this.updateView(t);
                });
        }
    }
    updateView(value) {
        if (this.viewRef) {
            this.viewContainerRef.clear();
            this.viewRef = null;
        }
        if (value)
            this.viewRef = this.viewContainerRef.createEmbeddedView(this.templateRef, this._context);
        else
            this.viewContainerRef.clear();
    }
};
tslib_1.__decorate([
    Input('rxAuthorize'),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], RxAuthorizeDirective.prototype, "component", null);
RxAuthorizeDirective = tslib_1.__decorate([
    Directive({
        selector: '[rxAuthorize]'
    }),
    tslib_1.__metadata("design:paramtypes", [ViewContainerRef, TemplateRef, Injector])
], RxAuthorizeDirective);
export { RxAuthorizeDirective };
