// @rxweb/http/esm5/models/http-request.js
import { isFormData, isBlob, isArrayBuffer } from "../functions/util";
import { createQueryPath } from "../functions/create-query-path";
var HttpRequest = /** @class */ (function () {
    function HttpRequest(body, headers, host, isHttps, method, path, params, queryParams, responseType, scheme, fullPath) {
        this.body = body;
        this.headers = headers;
        this.host = host;
        this.method = method;
        this.path = path;
        this.params = params;
        this.queryParams = queryParams;
        this.responseType = responseType;
        this.scheme = scheme;
        if (!responseType)
            this.responseType = 'json';
        this.isHttps = isHttps;
        if (fullPath)
            this.fullPath = fullPath;
    }
    Object.defineProperty(HttpRequest.prototype, "isHttps", {
        get: function () {
            return this._isHttps;
        },
        set: function (value) {
            this._isHttps = value;
            if (value)
                this.fullPath = this.fullPath.replace("http:", "https:");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HttpRequest.prototype, "fullPath", {
        get: function () {
            return this._fullPath ? this._fullPath : this.host + "/" + this.path + createQueryPath({ params: this.params, queryParams: this.queryParams });
        },
        set: function (value) {
            this._fullPath = value;
        },
        enumerable: true,
        configurable: true
    });
    HttpRequest.prototype.getContentType = function () {
        if (this.body === null) {
            return null;
        }
        if (isFormData(this.body)) {
            return null;
        }
        if (isBlob(this.body)) {
            return this.body.type || null;
        }
        if (isArrayBuffer(this.body)) {
            return null;
        }
        if (typeof this.body === 'string') {
            return 'text/plain';
        }
        if (typeof this.body === 'object' || typeof this.body === 'number' ||
            Array.isArray(this.body)) {
            return 'application/json';
        }
        return null;
    };
    HttpRequest.prototype.serializeBody = function () {
        if (this.body === null) {
            return null;
        }
        if (isArrayBuffer(this.body) || isBlob(this.body) || isFormData(this.body) ||
            typeof this.body === 'string') {
            return this.body;
        }
        if (this.method == 'PATCH')
            this.body = this.convertPatchBody(this.body);
        if (typeof this.body === 'object' || typeof this.body === 'boolean' ||
            Array.isArray(this.body)) {
            return JSON.stringify(this.body);
        }
        if (this.body)
            return String(this.body);
        return '';
    };
    HttpRequest.prototype.convertPatchBody = function (object) {
        var patchBody = [];
        for (var col in object) {
            patchBody.push({
                'op': 'replace',
                'path': '/' + col,
                'value': object[col]
            });
        }
        return patchBody;
    };
    return HttpRequest;
}());
export { HttpRequest };
