// @rxweb/angular-router/esm2015/core/core-component.js
import { RxHttp } from "@rxweb/http";
import { ClientDataStorage } from "@rxweb/storage";
import { componentInstanceProvider } from "./component-instance-provider";
import { routeContainer } from "./route-container";
import { encoder } from "../core/encoder";
export class CoreComponent extends RxHttp {
    onQueryParamsChanged() { }
    onParamsChanged() { }
    constructor() {
        super();
        let result = componentInstanceProvider.getResult();
        if (result && result.activatedRouteSnapshot) {
            this.params = result.activatedRouteSnapshot.params;
            this.queryParams = result.activatedRouteSnapshot.queryParams;
            this.navigation = result.router;
            this.user = routeContainer.getUser();
            this.data = result.activatedRouteSnapshot.data;
            this.storage = new ClientDataStorage();
            this.parentAuthorizeConfig = result.authorizeConfig;
            let decoratorConfig = routeContainer.getModelDecorator(this.constructor, "access");
            if (decoratorConfig)
                this.authorizeConfig = decoratorConfig.functions;
            componentInstanceProvider.add(this.constructor.name, this.instanceProvider.bind(this), this.paramsChange.bind(this), this.queryParamsChange.bind(this));
        }
    }
    instanceProvider() {
        return this;
    }
    resolve(component) {
        return componentInstanceProvider.resolve(component.constructor);
    }
    navigate(commands, extras) {
        return this.navigation.navigate(commands.concat(this.getParsedParams(extras)), extras);
    }
    ngOnDestroy() {
        componentInstanceProvider.dispose(this.constructor);
    }
    paramsChange(params) {
        this.params = params;
        this.onParamsChanged();
    }
    queryParamsChange(queryParams) {
        this.queryParams = queryParams;
        this.onQueryParamsChanged();
    }
    getParsedParams(extras) {
        let params = [];
        if (extras && extras.params) {
            var isEncryption = routeContainer.get().urlEncryption;
            extras.params.forEach(t => {
                let encode = isEncryption ? encoder.encode(t) : t;
                params.push(encode);
            });
        }
        return params;
    }
}
