// @rxweb/angular-router/esm5/directives/rx-authorize.directive.js
import * as tslib_1 from "tslib";
import { Injector, Directive, ViewContainerRef, TemplateRef, Input } from "@angular/core";
import { NgIfContext } from "@angular/common";
import { routeContainer } from "../core";
import { componentInstanceProvider } from "../core/component-instance-provider";
var RxAuthorizeDirective = /** @class */ (function () {
    function RxAuthorizeDirective(viewContainerRef, templateRef, injector) {
        this.viewContainerRef = viewContainerRef;
        this.templateRef = templateRef;
        this.injector = injector;
        this.viewRef = null;
        this._context = new NgIfContext();
    }
    Object.defineProperty(RxAuthorizeDirective.prototype, "component", {
        set: function (value) {
            if (Array.isArray(value)) {
                var result = true;
                this._components = value;
                this.checkAuth(0);
            }
            else if (value)
                this.checkAccess(value);
        },
        enumerable: true,
        configurable: true
    });
    RxAuthorizeDirective.prototype.checkAuth = function (index) {
        if (this._components.length > index) {
            this.checkAccess(this._components[index], index);
        }
        else {
            this.updateView(true);
        }
    };
    RxAuthorizeDirective.prototype.checkAccess = function (value, currentIndex) {
        var _this = this;
        if (currentIndex === void 0) { currentIndex = 0; }
        var authorizeModel = routeContainer.get().authorization;
        var component = routeContainer.getModelDecorator(value, "access");
        if (authorizeModel && component) {
            var authorize = this.injector.get(authorizeModel);
            var authorizeConfig = componentInstanceProvider.getAuthorizeConfig();
            var result = authorize.authorizeChildren(component.functions, authorizeConfig);
            if (typeof result === "boolean") {
                if (this._components && this._components.length > 0 && result) {
                    var index = currentIndex + 1;
                    this.checkAuth(index);
                }
                else
                    this.updateView(result);
            }
            else
                result.then(function (t) {
                    if (t) {
                        var index = currentIndex + 1;
                        _this.checkAuth(index);
                    }
                    else
                        _this.updateView(t);
                });
        }
    };
    RxAuthorizeDirective.prototype.updateView = function (value) {
        if (this.viewRef) {
            this.viewContainerRef.clear();
            this.viewRef = null;
        }
        if (value)
            this.viewRef = this.viewContainerRef.createEmbeddedView(this.templateRef, this._context);
        else
            this.viewContainerRef.clear();
    };
    tslib_1.__decorate([
        Input('rxAuthorize'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], RxAuthorizeDirective.prototype, "component", null);
    RxAuthorizeDirective = tslib_1.__decorate([
        Directive({
            selector: '[rxAuthorize]'
        }),
        tslib_1.__metadata("design:paramtypes", [ViewContainerRef, TemplateRef, Injector])
    ], RxAuthorizeDirective);
    return RxAuthorizeDirective;
}());
export { RxAuthorizeDirective };
