// @rxweb/angular-router/esm5/directives/rx-multilingual.directive.js
import * as tslib_1 from "tslib";
import { Injector, Directive, ViewContainerRef, TemplateRef, Input } from "@angular/core";
import { NgIfContext } from "@angular/common";
import { routeContainer } from "../core";
var RxMultilingualDirective = /** @class */ (function () {
    function RxMultilingualDirective(viewContainerRef, templateRef, injector) {
        this.viewContainerRef = viewContainerRef;
        this.templateRef = templateRef;
        this.injector = injector;
        this.viewRef = null;
        this._context = new NgIfContext();
    }
    Object.defineProperty(RxMultilingualDirective.prototype, "component", {
        set: function (value) {
            var _this = this;
            var childMultiLingual = routeContainer.get().childMultilingualResolver;
            if (childMultiLingual) {
                var childMultilingual = this.injector.get(childMultiLingual);
                var result = childMultilingual.resolveChildren(this.viewContainerRef["_view"].component.constructor);
                if (typeof result == "boolean")
                    this.updateView(result);
                else
                    result.then(function (t) {
                        _this.updateView(t);
                    });
            }
        },
        enumerable: true,
        configurable: true
    });
    RxMultilingualDirective.prototype.updateView = function (value) {
        if (this.viewRef) {
            this.viewContainerRef.clear();
            this.viewRef = null;
        }
        if (value)
            this.viewRef = this.viewContainerRef.createEmbeddedView(this.templateRef, this._context);
        else
            this.viewContainerRef.clear();
    };
    tslib_1.__decorate([
        Input('rxMultilingual'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], RxMultilingualDirective.prototype, "component", null);
    RxMultilingualDirective = tslib_1.__decorate([
        Directive({
            selector: '[rxMultilingual]'
        }),
        tslib_1.__metadata("design:paramtypes", [ViewContainerRef, TemplateRef, Injector])
    ], RxMultilingualDirective);
    return RxMultilingualDirective;
}());
export { RxMultilingualDirective };
