// @rxweb/http/esm5/domain/rx-http-response.js
import * as tslib_1 from "tslib";
import { XhrRequest } from "./xhr-request";
import { createInstance } from "../functions/create-instance";
import { HttpResponse, XhrContext } from '../models';
var RxHttpResponse = /** @class */ (function () {
    function RxHttpResponse(request, filters, inMemoryFilter, onError, badRequest) {
        var _this = this;
        this.request = request;
        this.filters = filters;
        this.inMemoryFilter = inMemoryFilter;
        this.onError = onError;
        this.badRequest = badRequest;
        this._index = 0;
        this._onResponse = new Array();
        this.onRequestExecuting = function () {
            return function (context) {
                if (context.response.statusCode) {
                    var response = tslib_1.__assign({}, context.response, { body: context.result });
                    if (response.statusCode >= 200 && response.statusCode < 300)
                        _this._onSuccess(response);
                    else {
                        _this._onError.error(response);
                        _this.onError(response);
                    }
                }
                else
                    _this.executeFilterOnRequest();
            };
        };
    }
    RxHttpResponse.prototype.process = function (type, success) {
        this._onSuccess = this.onComplete(type, success);
        this._onError = success;
        this.inMemoryFilter ? this.callFilter(this.inMemoryFilter) : this.executeFilterOnRequest();
    };
    RxHttpResponse.prototype.executeFilterOnRequest = function () {
        if (this.filters && this.filters.length > 0 && this.filters.length > this._index) {
            this.callFilter(this.filters[this._index]);
        }
        else
            new XhrRequest(this.request, this._onSuccess, this.onRequestError.bind(this), this.badRequest);
    };
    RxHttpResponse.prototype.callFilter = function (filterConfig) {
        if (!filterConfig.runsOn || filterConfig.runsOn.indexOf(this.request.method) != -1) {
            var modelInstance = createInstance(filterConfig.model, filterConfig.args || []);
            modelInstance.onRequestExecuting = this.onRequestExecuting();
            this._index++;
            modelInstance.onRequest(new XhrContext(this.request, new HttpResponse()));
            if (modelInstance["onResponse"])
                this._onResponse.push(modelInstance);
        }
        else
            this._index++;
    };
    RxHttpResponse.prototype.onComplete = function (resultType, onSuccess) {
        var _this = this;
        return function (response) {
            if (_this._onResponse.length > 0 && !response.isInMemoryResponse)
                _this._onResponse.forEach(function (t) { return t.onResponse(response); });
            onSuccess.next(response.body);
            onSuccess.complete();
        };
    };
    RxHttpResponse.prototype.onRequestError = function (response) {
        this._onError.error(response);
        if (this.onError)
            this.onError(response);
    };
    return RxHttpResponse;
}());
export { RxHttpResponse };
