// @rxweb/http/esm2015/domain/rx-http-response.js
import { XhrRequest } from "./xhr-request";
import { createInstance } from "../functions/create-instance";
import { HttpResponse, XhrContext } from '../models';
export class RxHttpResponse {
    constructor(request, filters, inMemoryFilter, onError, badRequest) {
        this.request = request;
        this.filters = filters;
        this.inMemoryFilter = inMemoryFilter;
        this.onError = onError;
        this.badRequest = badRequest;
        this._index = 0;
        this._onResponse = new Array();
        this.onRequestExecuting = () => {
            return (context) => {
                if (context.response.statusCode) {
                    let response = Object.assign({}, context.response, { body: context.result });
                    if (response.statusCode >= 200 && response.statusCode < 300)
                        this._onSuccess(response);
                    else {
                        this._onError.error(response);
                        this.onError(response);
                    }
                }
                else
                    this.executeFilterOnRequest();
            };
        };
    }
    process(type, success) {
        this._onSuccess = this.onComplete(type, success);
        this._onError = success;
        this.inMemoryFilter ? this.callFilter(this.inMemoryFilter) : this.executeFilterOnRequest();
    }
    executeFilterOnRequest() {
        if (this.filters && this.filters.length > 0 && this.filters.length > this._index) {
            this.callFilter(this.filters[this._index]);
        }
        else
            new XhrRequest(this.request, this._onSuccess, this.onRequestError.bind(this), this.badRequest);
    }
    callFilter(filterConfig) {
        if (!filterConfig.runsOn || filterConfig.runsOn.indexOf(this.request.method) != -1) {
            let modelInstance = createInstance(filterConfig.model, filterConfig.args || []);
            modelInstance.onRequestExecuting = this.onRequestExecuting();
            this._index++;
            modelInstance.onRequest(new XhrContext(this.request, new HttpResponse()));
            if (modelInstance["onResponse"])
                this._onResponse.push(modelInstance);
        }
        else
            this._index++;
    }
    onComplete(resultType, onSuccess) {
        return (response) => {
            if (this._onResponse.length > 0 && !response.isInMemoryResponse)
                this._onResponse.forEach(t => t.onResponse(response));
            onSuccess.next(response.body);
            onSuccess.complete();
        };
    }
    onRequestError(response) {
        this._onError.error(response);
        if (this.onError)
            this.onError(response);
    }
}
