// @rxweb/angular-router/esm5/core/component-instance-provider.js
export var componentInstanceProvider = new (/** @class */ (function () {
    function class_1() {
        this.instances = new Array();
    }
    class_1.prototype.broadCastParamChange = function (params) {
        if (Object.keys(params).length > 0)
            this.instances.forEach(function (t) {
                t.paramChanged(params);
            });
    };
    class_1.prototype.broadCastQueryParamChange = function (queryParams) {
        if (Object.keys(queryParams).length > 0)
            this.instances.forEach(function (t) {
                t.queryParamChanged(queryParams);
            });
    };
    class_1.prototype.setComponentRoute = function (route, router) {
        this.activatedRouteSnapshot = route;
        this.router = router;
        this.broadCastParamChange(route.params);
        this.broadCastQueryParamChange(route.queryParams);
    };
    class_1.prototype.setAuthroizeConfig = function (config) {
        this.authorizeConfig = config;
    };
    class_1.prototype.getAuthorizeConfig = function () {
        return this.authorizeConfig;
    };
    class_1.prototype.getResult = function () {
        return {
            activatedRouteSnapshot: this.activatedRouteSnapshot,
            router: this.router,
            authorizeConfig: this.authorizeConfig
        };
    };
    class_1.prototype.getActivatedRouteSnapshot = function () {
        return this.activatedRouteSnapshot;
    };
    class_1.prototype.getRouter = function () {
        return this.router;
    };
    class_1.prototype.add = function (name, resolver, paramChanged, queryParamChanged) {
        this.instances.push({ name: name, resolver: resolver, paramChanged: paramChanged, queryParamChanged: queryParamChanged });
    };
    class_1.prototype.resolve = function (target) {
        var instance = this.instances.filter(function (t) { return t.name == target.name; })[0];
        if (instance)
            return instance.resolver();
        throw new Error("Component instance is not available");
    };
    class_1.prototype.dispose = function (target) {
        var instances = this.instances.filter(function (t) { return t.name == target.name; });
        var length = instances.length;
        for (var i = 0; i < length; i++)
            instances.splice(0, 1);
    };
    return class_1;
}()))();
