// @rxweb/angular-router/esm5/core/core-component.js
import * as tslib_1 from "tslib";
import { RxHttp } from "@rxweb/http";
import { ClientDataStorage } from "@rxweb/storage";
import { componentInstanceProvider } from "./component-instance-provider";
import { routeContainer } from "./route-container";
import { encoder } from "../core/encoder";
var CoreComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CoreComponent, _super);
    function CoreComponent() {
        var _this = _super.call(this) || this;
        var result = componentInstanceProvider.getResult();
        if (result && result.activatedRouteSnapshot) {
            _this.params = result.activatedRouteSnapshot.params;
            _this.queryParams = result.activatedRouteSnapshot.queryParams;
            _this.navigation = result.router;
            _this.user = routeContainer.getUser();
            _this.data = result.activatedRouteSnapshot.data;
            _this.storage = new ClientDataStorage();
            _this.parentAuthorizeConfig = result.authorizeConfig;
            var decoratorConfig = routeContainer.getModelDecorator(_this.constructor, "access");
            if (decoratorConfig)
                _this.authorizeConfig = decoratorConfig.functions;
            componentInstanceProvider.add(_this.constructor.name, _this.instanceProvider.bind(_this), _this.paramsChange.bind(_this), _this.queryParamsChange.bind(_this));
        }
        return _this;
    }
    CoreComponent.prototype.onQueryParamsChanged = function () { };
    CoreComponent.prototype.onParamsChanged = function () { };
    CoreComponent.prototype.instanceProvider = function () {
        return this;
    };
    CoreComponent.prototype.resolve = function (component) {
        return componentInstanceProvider.resolve(component.constructor);
    };
    CoreComponent.prototype.navigate = function (commands, extras) {
        return this.navigation.navigate(commands.concat(this.getParsedParams(extras)), extras);
    };
    CoreComponent.prototype.ngOnDestroy = function () {
        componentInstanceProvider.dispose(this.constructor);
    };
    CoreComponent.prototype.paramsChange = function (params) {
        this.params = params;
        this.onParamsChanged();
    };
    CoreComponent.prototype.queryParamsChange = function (queryParams) {
        this.queryParams = queryParams;
        this.onQueryParamsChanged();
    };
    CoreComponent.prototype.getParsedParams = function (extras) {
        var params = [];
        if (extras && extras.params) {
            var isEncryption = routeContainer.get().urlEncryption;
            extras.params.forEach(function (t) {
                var encode = isEncryption ? encoder.encode(t) : t;
                params.push(encode);
            });
        }
        return params;
    };
    return CoreComponent;
}(RxHttp));
export { CoreComponent };
